### PR TITLE
Basic megaflow implementation

### PR DESCRIPTION
**Pull request by YiHungWei**

### CODE REVIEW

File: bpf/lookup.h [hunk 1]
```diff
@@ -154,10 +154,69 @@ flow_stats_account(struct ebpf_headers_t *headers,
...
+    int i;
+
+    #pragma unroll
+    for (i = 0; i < sizeof *dst; i += sizeof(long)) {
+        *d++ = *s++ & *m++;
+    }
+}
+
+static inline struct bpf_action_batch *
+megaflow_lookup(struct bpf_flow_key *key OVS_UNUSED)
```
Comment (yifsun) [inline bpf/lookup.h:173]: maybe remove OVS_UNUSED if it not useful.

Comment (YiHungWei) [reply bpf/lookup.h:173]: Sure, I will remove that in the next version. Thanks

File: lib/dpif-bpf.c [hunk 2]
```diff
@@ -1108,31 +1110,293 @@ fetch_flow(struct dpif_bpf_dp *dp, struct dpif_flow *flow,
...
+                        return -1;
+                    }
+                    dpif_bpf_update_mask_refcount(i, false);
+                    return 0;
+                } else {
+                    VLOG_ERR("Failed to delete megaflow. Same masked key but "
+                             "different key is observed.");
+                }
+            } else {
+            if (err) {
```
Comment (yifsun) [inline lib/dpif-bpf.c:1382]: there seems an indent issue.

File: lib/dpif-bpf.c [hunk 3]
```diff
@@ -1108,31 +1110,293 @@ fetch_flow(struct dpif_bpf_dp *dp, struct dpif_flow *flow,
...
+    new_megaflow_entry.action_batch = *actions;
+
+    /* Found existing mask in the mask table */
+    if (ret == 1) {
+        /* Found existing megaflow in the megaflow table */
+        if (bpf_map_lookup_elem(datapath.bpf.megaflow_table.fd,
+                                &megaflow_key, &old_megaflow_entry) == 0) {
+            VLOG_INFO("Update action to an existing megaflow.");
+
+            ovs_assert(datapath.bpf.megaflow_table.fd != -1);
```
Comment (yifsun) [inline lib/dpif-bpf.c:1304]: I am not sure what is the reason for assert here, can you explain?

File: lib/dpif-bpf.c [hunk 4]
```diff
@@ -1108,31 +1110,293 @@ fetch_flow(struct dpif_bpf_dp *dp, struct dpif_flow *flow,
...
+                return -1;
+            }
+            ret = bpf_map_update_elem(datapath.bpf.megaflow_table.fd,
+                                      &new_megaflow_entry, actions, BPF_ANY);
+            if (ret) {
+                VLOG_ERR("Failed to add flow into flow table, map fd %d, "
+                         "error %s",
+                         datapath.bpf.flow_table.fd, ovs_strerror(errno));
+                return errno;
+            }
```
Comment (yifsun) [inline lib/dpif-bpf.c:1317]: line 1310-1317 look redundant with code below.

File: bpf/lookup.h [hunk 5]
```diff
@@ -154,10 +154,69 @@ flow_stats_account(struct ebpf_headers_t *headers,
...
+    const long *m = (const long *) mask;
+    const long *s = (const long *) src;
+    long *d = (long *) dst;
+    int i;
+
+    #pragma unroll
+    for (i = 0; i < sizeof *dst; i += sizeof(long)) {
+        *d++ = *s++ & *m++;
+    }
+}
```
Comment (yifsun) [inline bpf/lookup.h:170]: do we assert size of bpf_flow_key is exactly sizeof(long)*n bytes?

File: lib/dpif-bpf.c (hunk 2 above)
Comment (YiHungWei) [reply lib/dpif-bpf.c:1382]: Yes, will address in the next version.

File: lib/dpif-bpf.c (hunk 3 above)
Comment (YiHungWei) [reply lib/dpif-bpf.c:1304]: Yes, I found that I can combine the two ovs_assert() in one place, updated in v2.

File: lib/dpif-bpf.c (hunk 4 above)
Comment (YiHungWei) [reply lib/dpif-bpf.c:1317]: There are 3 cases here, and the logging and ref_count update will be different. 
1) mask exist, megaflow does not exist
2) mask exist, megaflow does exist
3) mask is not exist

case 1) is a special case, and case 2) and 3) share the common code path. 

The current design reflects how we deal with the 3 cases.

File: bpf/lookup.h (hunk 5 above)
Comment (YiHungWei) [reply bpf/lookup.h:170]: bpf_flow_key is long aligned.

struct bpf_flow_key {
    struct ebpf_headers_t headers;
    struct ebpf_metadata_t mds;
} __attribute__((aligned(sizeof(long))));

File: lib/dpif-bpf.c (hunk 4 above)
Comment (yifsun) [reply lib/dpif-bpf.c:1317]: I see.
In the bpf_map_update_elem above, the value is actions, is it a supposed value? I saw that value of megaflow_table is of type mega_flow_entry.

File: bpf/lookup.h (hunk 5 above)
Comment (yifsun) [reply bpf/lookup.h:170]: Maybe I miss something. attribute(aligned(sizeof(long))) means the memory will be allocated at sizeof(long) boundary. Does that mean that memory block contains all sizeof(long)*n bytes?

for example,
struct temp {
    char data[1];
} attribute ((aligned(sizeof(long)));

is sizeof(struct temp) 1 or sizeof(long)?

**Comment by YiHungWei** [comment]
I sent out a v2 in the other pull request that addresses Yifeng's comment.  Will close this one.

File: lib/dpif-bpf.c (hunk 4 above)
Comment (YiHungWei) [reply lib/dpif-bpf.c:1317]: Good catch.  It should be megaflow_entry. Thanks!

File: bpf/lookup.h (hunk 5 above)
Comment (YiHungWei) [reply bpf/lookup.h:170]: struct temp will be size of long, at least in gcc 5.4.0

```#include <stdio.h>

int main()
{
    struct temp {
        char data[1];
    } __attribute__((  aligned(sizeof(long))) );

    printf("%lu\n", sizeof(struct temp));

    return 0;
}
```

Comment (yifsun) [reply bpf/lookup.h:170]: thanks! that is good to know.

**Pull request closed without merging**